apax/train/trainer.py

```diff
@@ -156,38 +158,76 @@ def calc_loss(params, inputs, labels, loss_fn, model):
     return loss, predictions
 
 
-def make_step_fns(loss_fn, Metrics, model, sam_rho):
+def make_ensemble_update(update_fn: Callable) -> Callable:
+    v_update_fn = jax.vmap(update_fn, (0, None, None), (0, 0, 0))
+
+    def ensemble_update_fn(state, inputs, labels):
+        loss, predictions, state = v_update_fn(state, inputs, labels)
+
+        mean_predictions = jax.tree_map(lambda x: jnp.mean(x, axis=0), predictions)
+        mean_loss = jnp.mean(loss)
+        # TODO Add std to predictions
+        return mean_loss, mean_predictions, state
```

Comment:
could this function be moved outside and returned with `partial(ensemble_update_fn, v_update_fn=v_update_fn)` to reduce code cuplication (see `make_ensemble_eval`) ?

Reply:
The only part that is really duplicated is
```
mean_predictions = jax.tree_map(lambda x: jnp.mean(x, axis=0), predictions)
mean_loss = jnp.mean(loss)
```

I think 2 lines are acceptable

Reply:
I guess I could unify them by using tuple (un)packing, but I'm not sure if this helps readability.
We should however adopt proper type hints. I'll make an issue.